File: backends/php-laravel/app/UseCases/Gate/GateService.php
<?php

/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */

declare(strict_types=1);

namespace App\UseCases\Gate;

use App\Domain\Client\Condition\ClientCondition;
use App\Domain\Client\Repositories\ClientRepository;
use App\Domain\Gate\Repositories\GateRepository;
use App\Domain\Gate\ValueObjects\GateIssueVo;
use App\Domain\Gate\ValueObjects\GateVerifyVo;
use App\Domain\Client\Entities\JwtHistory;
use App\Domain\Client\Repositories\JwtHistoryRepository;
use App\Support\Exceptions\AppException;
use App\Support\Mappers\SimpleMapper;
use App\Support\Services\AbstractService;
use App\UseCases\Gate\Dtos\GateIssueDto;
use App\UseCases\Gate\Dtos\GateVerifyDto;
use Carbon\Carbon;
use Firebase\JWT\JWT;
use Firebase\JWT\Key;
use Illuminate\Support\Str;
use Throwable;

/**
 * 認可Serviceクラスです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 * @package App\UseCases\Gate
 */
class GateService extends AbstractService
{
    /**
     * コンストラクタ。
     *
     * @param ClientRepository $clientRepository クライアントリポジトリ
     * @param GateRepository $gateRepository 認可リポジトリ
     * @param JwtHistoryRepository $historyRepository JWT履歴リポジトリ
     */
    public function __construct(
        private readonly ClientRepository $clientRepository,
        private readonly GateRepository $gateRepository,
        private readonly JwtHistoryRepository $historyRepository,
    ) {
    }

    /**
     * JWTを発行します。
     *
     * @param GateIssueDto $dto JWT発行DTO
     * @return GateIssueVo JWT 発行結果 ValueObject
     * @throws \AutoMapperPlus\Exception\UnregisteredMappingException マッピング例外
     */
    public function issueToken(GateIssueDto $dto): GateIssueVo
    {
        $condition = SimpleMapper::map($dto, ClientCondition::class);
        $client = $this->clientRepository->findByAccessToken($condition);
        if (is_null($client)) {
            throw AppException::unauthorized('client_not_found');
        }

        /** @var array{issuer: string, algorithm: string, ttl: int, cache_ttl: int} $configs */
        $configs = config('authorization.app.jwt');
        $identifier = (string)$client->identifier;
        $token = $this->gateRepository->getJwt($identifier, $dto->memberId);

        if ($token === null) {
            $token = $this->issueJwt(
                $configs,
                $dto->memberId,
                $identifier,
                (string)$client->privateKey,
                (string)$client->fingerprint,
            );
            $this->gateRepository->putJwt($identifier, $dto->memberId, $token, $configs['cache_ttl']);

            $history = new JwtHistory();
            $history->clientId = $client->id;
            $history->memberId = $dto->memberId;
            $history->issueAt = Carbon::now();
            $history->jwt = $token;
            $history->assignCreated(0);
            $this->historyRepository->persist($history);
        }

        return new GateIssueVo()->assign(['token' => $token]);
    }

    /**
     * JWTを検証します。
     *
     * @param GateVerifyDto $dto JWT 検証リクエスト用 DTO
     * @return GateVerifyVo JWT 検証結果（Payload 相当）ValueObject
     * @throws \AutoMapperPlus\Exception\UnregisteredMappingException マッピング例外
     */
    public function verify(GateVerifyDto $dto): GateVerifyVo
    {
        $condition = SimpleMapper::map($dto, ClientCondition::class);
        $client = $this->clientRepository->findByIdentifier($condition);
        if (is_null($client)) {
            throw AppException::forbidden('client_not_found');
        }

        $payload = $this->verifyJwt(
            $dto->identifier,
            $dto->token,
            (string)$client->publicKey,
        );

        $vo = new GateVerifyVo();
        $vo->assign($payload);

        return $vo;
    }

    /**
     * RS256 で署名した JWT を発行します。
     *
     * @param array{issuer: string, algorithm: string, ttl: int} $configs JWT 設定
     * @param string $memberId クライアント会員ID（sub）
     * @param string $identifier クライアント識別名（aud）
     * @param string $privateKey 署名用 RSA 秘密鍵（PEM 形式）
     * @param string $fingerprint 秘密鍵フィンガープリント（kid）
     * @return string 発行した JWT 文字列
     */
    private function issueJwt(
        array $configs,
        string $memberId,
        string $identifier,
        string $privateKey,
        string $fingerprint
    ): string {
        $now = time();
        $payload = [
            'iss' => $configs['issuer'],
            'sub' => $memberId,
            'aud' => $identifier,
            'exp' => $now + $configs['ttl'],
            'iat' => $now,
            'nbf' => $now,
            'jti' => (string)Str::uuid(),
        ];

        return JWT::encode($payload, $privateKey, $configs['algorithm'], $fingerprint);
    }

    /**
     * JWT を検証しデコードした Payload を返します。
     *
     * @param string $identifier クライアント識別名（aud 検証に使用）
     * @param string $token JWT 文字列
     * @param string $publicKey 検証用 RSA 公開鍵（PEM 形式）
     * @return array<string, mixed> デコードされた Payload
     * @throws AppException JWT が無効な場合
     */
    private function verifyJwt(string $identifier, string $token, string $publicKey): array
    {
        /** @var array{issuer: string, algorithm: string} $jwt */
        $jwt = config('authorization.app.jwt');

        try {
            $decoded = JWT::decode($token, new Key($publicKey, $jwt['algorithm']));
        } catch (Throwable) {
            throw AppException::unauthorized('jwt_invalid');
        }

        $payload = (array)$decoded;

        if (($payload['iss'] ?? '') !== $jwt['issuer']) {
            throw AppException::unauthorized('jwt_invalid');
        }

        if (($payload['aud'] ?? '') !== $identifier) {
            throw AppException::forbidden('jwt_invalid');
        }

        return $payload;
    }
}

File: backends/java-spring/src/main/java/com/authorization/usecases/gate/GateService.java
/**
 * This is a program developed by BobTabo.
 *
 * Copyright (c) 2026 BobTabo. All Rights Reserved.
 */
package com.authorization.usecases.gate;

import com.authorization.config.AppConfig;
import com.authorization.domain.client.condition.ClientCondition;
import com.authorization.domain.client.entities.Client;
import com.authorization.domain.client.entities.JwtHistory;
import com.authorization.domain.client.repositories.ClientRepository;
import com.authorization.domain.client.repositories.JwtHistoryRepository;
import com.authorization.domain.gate.repositories.GateRepository;
import com.authorization.domain.gate.valueobjects.GateIssueVo;
import com.authorization.domain.gate.valueobjects.GateVerifyVo;
import com.authorization.support.exceptions.AppException;
import com.authorization.support.services.AbstractService;
import com.authorization.usecases.gate.dtos.GateIssueDto;
import com.authorization.usecases.gate.dtos.GateVerifyDto;
import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.crypto.RSASSASigner;
import com.nimbusds.jose.crypto.RSASSAVerifier;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.text.ParseException;
import java.time.Instant;
import java.time.LocalDateTime;
import java.util.Base64;
import java.util.Date;
import java.util.UUID;

/**
 * 認可Serviceクラスです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
public class GateService extends AbstractService {

    private final ClientRepository clientRepository;
    private final GateRepository gateRepository;
    private final JwtHistoryRepository historyRepository;
    private final AppConfig.Jwt jwtConfig;

    /**
     * コンストラクタ。
     *
     * @param clientRepository クライアントRepository
     * @param gateRepository 認可Repository
     * @param historyRepository JWT履歴Repository
     * @param jwtConfig JWT設定
     */
    public GateService(
            ClientRepository clientRepository,
            GateRepository gateRepository,
            JwtHistoryRepository historyRepository,
            AppConfig.Jwt jwtConfig) {
        this.clientRepository = clientRepository;
        this.gateRepository = gateRepository;
        this.historyRepository = historyRepository;
        this.jwtConfig = jwtConfig;
    }

    /**
     * JWTを発行します。
     *
     * @param dto JWT発行DTO
     * @return JWT発行結果ValueObject
     */
    public GateIssueVo issueToken(GateIssueDto dto) {
        ClientCondition condition = new ClientCondition();
        condition.setAccessToken(dto.getAccessToken());
        Client client = clientRepository.findByAccessToken(condition);
        if (client == null) {
            throw AppException.unauthorized("client_not_found");
        }

        String identifier = client.getIdentifier();
        String token = gateRepository.getJwt(identifier, dto.getMemberId());

        if (token == null) {
            token = issueJwt(dto.getMemberId(), identifier, client.getPrivateKey(), client.getFingerprint());
            gateRepository.putJwt(identifier, dto.getMemberId(), token, (int) jwtConfig.cacheTtl());

            JwtHistory history = new JwtHistory();
            history.setClientId(client.getId());
            history.setMemberId(dto.getMemberId());
            history.setIssueAt(LocalDateTime.now());
            history.setJwt(token);
            history.assignCreated(0);
            historyRepository.persist(history);
        }

        GateIssueVo vo = new GateIssueVo();
        vo.setToken(token);
        return vo;
    }

    /**
     * JWTを検証します。
     *
     * @param dto JWT検証リクエスト用DTO
     * @return JWT検証結果ValueObject
     */
    public GateVerifyVo verify(GateVerifyDto dto) {
        ClientCondition condition = new ClientCondition();
        condition.setIdentifier(dto.getIdentifier());
        Client client = clientRepository.findByIdentifier(condition);
        if (client == null) {
            throw AppException.forbidden("client_not_found");
        }

        return verifyJwt(dto.getIdentifier(), dto.getToken(), client.getPublicKey());
    }

    /**
     * RS256 で署名した JWT を発行します。
     *
     * @param memberId クライアント会員ID（sub）
     * @param identifier クライアント識別名（aud）
     * @param privateKeyPem 署名用RSA秘密鍵（PEM形式）
     * @param fingerprint 秘密鍵フィンガープリント（kid）
     * @return 発行したJWT文字列
     */
    private String issueJwt(String memberId, String identifier, String privateKeyPem, String fingerprint) {
        try {
            RSAPrivateKey privateKey = loadPrivateKey(privateKeyPem);
            Instant now = Instant.now();
            JWTClaimsSet claims = new JWTClaimsSet.Builder()
                    .issuer(jwtConfig.issuer())
                    .subject(memberId)
                    .audience(identifier)
                    .expirationTime(Date.from(now.plusSeconds(jwtConfig.ttl())))
                    .issueTime(Date.from(now))
                    .notBeforeTime(Date.from(now))
                    .jwtID(UUID.randomUUID().toString())
                    .build();
            JWSHeader header = new JWSHeader.Builder(JWSAlgorithm.RS256).keyID(fingerprint).build();
            SignedJWT jwt = new SignedJWT(header, claims);
            jwt.sign(new RSASSASigner(privateKey));
            return jwt.serialize();
        } catch (JOSEException | NoSuchAlgorithmException | InvalidKeySpecException e) {
            throw new IllegalStateException("failed to issue jwt", e);
        }
    }

    /**
     * JWT を検証しデコードした Payload を ValueObject で返します。
     *
     * @param identifier クライアント識別名（aud 検証に使用）
     * @param token JWT 文字列
     * @param publicKeyPem 検証用RSA公開鍵（PEM形式）
     * @return JWT検証結果ValueObject
     */
    private GateVerifyVo verifyJwt(String identifier, String token, String publicKeyPem) {
        JWTClaimsSet claims;
        try {
            RSAPublicKey publicKey = loadPublicKey(publicKeyPem);
            SignedJWT jwt = SignedJWT.parse(token);
            if (!jwt.verify(new RSASSAVerifier(publicKey))) {
                throw AppException.unauthorized("jwt_invalid");
            }
            claims = jwt.getJWTClaimsSet();
        } catch (ParseException | JOSEException | NoSuchAlgorithmException | InvalidKeySpecException e) {
            throw AppException.unauthorized("jwt_invalid");
        }

        if (claims.getExpirationTime() != null && new Date().after(claims.getExpirationTime())) {
            throw AppException.unauthorized("jwt_invalid");
        }
        if (!jwtConfig.issuer().equals(claims.getIssuer())) {
            throw AppException.unauthorized("jwt_invalid");
        }
        if (claims.getAudience() == null || !claims.getAudience().contains(identifier)) {
            throw AppException.forbidden("jwt_invalid");
        }

        GateVerifyVo vo = new GateVerifyVo();
        vo.setIss(claims.getIssuer());
        vo.setSub(claims.getSubject());
        vo.setAud(identifier);
        vo.setExp(claims.getExpirationTime() != null ? claims.getExpirationTime().getTime() / 1000 : 0);
        vo.setIat(claims.getIssueTime() != null ? claims.getIssueTime().getTime() / 1000 : 0);
        vo.setNbf(claims.getNotBeforeTime() != null ? claims.getNotBeforeTime().getTime() / 1000 : 0);
        vo.setJti(claims.getJWTID());
        return vo;
    }

    private static RSAPrivateKey loadPrivateKey(String pem) throws NoSuchAlgorithmException, InvalidKeySpecException {
        String cleaned = pem
                .replace("-----BEGIN PRIVATE KEY-----", "")
                .replace("-----END PRIVATE KEY-----", "")
                .replaceAll("\\s", "");
        byte[] bytes = Base64.getDecoder().decode(cleaned);
        return (RSAPrivateKey) KeyFactory.getInstance("RSA").generatePrivate(new PKCS8EncodedKeySpec(bytes));
    }

    private static RSAPublicKey loadPublicKey(String pem) throws NoSuchAlgorithmException, InvalidKeySpecException {
        String cleaned = pem
                .replace("-----BEGIN PUBLIC KEY-----", "")
                .replace("-----END PUBLIC KEY-----", "")
                .replaceAll("\\s", "");
        byte[] bytes = Base64.getDecoder().decode(cleaned);
        return (RSAPublicKey) KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(bytes));
    }
}
